### PR TITLE
Load SPC rom

### PR DESCRIPTION
**Pull request by melefo**
This pull request add :
 - A reworked & better APU Debugger
 - Separated Cartridge formats
 - Loading SPC Rom data to APU & DSP
 - Disable other components than APU if an Audio Rom is played
 - Fixed some opcodes

### CODE REVIEW

File: sources/APU/APU.cpp [hunk 1]
```diff
@@ -731,6 +825,74 @@ namespace ComSquare::APU
 			this->_renderer.playAudio(std::span{this->_soundBuffer}, samples / 2);
 	}
 
+	void APU::loadFromSPC(const std::shared_ptr<Cartridge::Cartridge>& cartridge)
+	{
+		const uint8_t *data = cartridge->getData();
+
+		std::string song = std::string(reinterpret_cast<const char *>(data + 0x2E), 0x20);
```
Comment (zoriya) [inline sources/APU/APU.cpp:835]: You don't seems to check the size of the cartrige, isn't that segfault prone?

File: sources/Cartridge/Cartridge.cpp [hunk 2]
```diff
@@ -156,10 +158,30 @@ namespace ComSquare::Cartridge
 		return bestAddress;
 	}
 
+	bool Cartridge::_isSPCFile()
+	{
+		std::string str = std::string(reinterpret_cast<char *>(this->_data), 0x21);
+
+		if (str != "SNES-SPC700 Sound File Data v0.30")
```
Comment (zoriya) [inline sources/Cartridge/Cartridge.cpp]: Is this string a magic string to confirm the header file?
If it is, you could make that a constexpr.

File: sources/Debugger/APUDebug.hpp [hunk 3]
```diff
@@ -11,8 +11,307 @@
...
+		int size;
+		std::tuple<Operand, Operand> operands;
+	};
+
+	class APUDebug : public APU::APU, public QObject
+	{
 	private:
+		//! @brief List of instructions and their information
+		std::array<Instruction, 0x100> _instructions {{
+          {"NOP", 1, {None, None}},
```
Comment (zoriya) [inline sources/Debugger/APUDebug.hpp]: Indent

File: sources/Cartridge/Cartridge.cpp [hunk 4]
```diff
@@ -156,10 +158,30 @@ namespace ComSquare::Cartridge
 		return bestAddress;
 	}
 
+	bool Cartridge::_isSPCFile()
+	{
+		std::string str = std::string(reinterpret_cast<char *>(this->_data), 0x21);
```
Comment (zoriya) [inline sources/Cartridge/Cartridge.cpp:165]: Check for cartridge size here too I think

**Review by zoriya — APPROVED** [review]

File: sources/APU/APU.cpp [hunk 5]
```diff
@@ -731,6 +825,77 @@ namespace ComSquare::APU
 			this->_renderer.playAudio(std::span{this->_soundBuffer}, samples / 2);
 	}
 
+	void APU::loadFromSPC(const std::shared_ptr<Cartridge::Cartridge>& cartridge)
+	{
+		const uint8_t *data = cartridge->getData();
+		uint24_t size = cartridge->getSize();
+		if (size < 0x101C0)
```
Comment (Octopus773) [inline sources/APU/APU.cpp:832]: if 0x101C0 have a meaning you should have a constexpr

File: sources/Cartridge/Cartridge.cpp [hunk 6]
```diff
@@ -98,13 +98,15 @@ namespace ComSquare::Cartridge
 
 	uint32_t Cartridge::_getHeaderAddress()
 	{
-		const std::vector<uint32_t> address = {0x7FC0, 0xFFC0, 0x81C0, 0x101C0};
+		const std::vector<uint32_t> address = {0x7FC0, 0xFFC0};
```
Comment (Octopus773) [inline sources/Cartridge/Cartridge.cpp:101]: same shit for hex values

**Review by Octopus773 — APPROVED** [review]

**Pull request merged**